README.md

```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
-
-When an addition message `c` is received:
-
-1. If there exists `r` in the rem-set such that `r.targetCastHash` equals `c.hash`, discard `c`
-2. Otherwise, add `c` into the add-set.
-
-When a remove message `r` is received:
+### 4.1.1 Add Messages
 
-1. If there is an `a` in the add-set where `a.hash` equals `d.targetCastHash`, delete it.
-2. If there is an `r` in the rem-set where `r.targetCastHash` equals `d.targetCastHash`
-   - If `r > d`, discard `d`
-   - If `r < d`, delete `r` and add `d` into the rem-set
-3. Otherwise, add `d` to the rem-set.
-
-### 4.1.2 Recasts
-
-A _Recast_ expresses an intent to share another cast. An account can recast any cast, including its own, but may not recast the same cast multiple times.
+A _Cast Add_ can contain up to 320 characters of unicode text and two URI's that can have up to 256 characters. Clients are responsible for unpacking and rendering the URI's along with the text. 
 
 ```ts
-type CastRecast = {
-  targetCastUri: URI;
-  schema: 'farcaster.xyz/schemas/v1/cast-recast';
+type CastAddMessage = {
+  embed: {
+    items: URI[];
+  },
+  parent?: URI;
+  text: string;
 };
 ```
 
-#### Message Validation
+A cast without a `parent` is a top-level cast, which clients should display on the user's profile or timeline. A cast with a `parent` is a reply to another cast, web URL or on-chain object which should be displayed in a thread.
 
-1. `schema` must be known.
-2. `targetCastUri` must be a valid Farcaster Cast URI and must not reference this message.
+Casts form a series of trees where each root is a Cast or URI and each child node is a reply cast. Each tree can be rendered as a thread in the UI. Trees are guaranteed to be acyclic because a parent must be hashed and signed before a child can point to it. A parent that changes its parent to point to a child will change its hash and break its relationship with the child.
 
-#### Set Construction
+```mermaid
+graph TB 
+    A((1))-->B((2))
+    A-->C((3))
+    A-->D((4))
+    B-->E((5))
+    B-->F((6))
+    B-->G((7))
+    C-->H((8))
+```
 
-Recasts can be stored using a two-phase set just like Short Text Casts. The remove operation is identical, while the addition operation is performed with `CastRecast`, and has one additional rule:
+A Cast message must pass the following validation steps:
 
-1. If there is a CastRecast `a` in the add **add-set** with the same `targetCastUri` and `fid` as the incoming CastRecast `rc`:
-   - If `a > rc`, discard `d`
-   - If `a < rc`, delete `a` and add `rc` into the rem-set
+1. `text` must contain <= 320 valid unicode characters
+2. `embed` must contain between 0 and 2 `items`
+3. `item` must be a URI of at most 256 characters
+4. `parent`, if present, must be a valid URI not equal to this message's URI
```

Comment:
Worth adding a mention of what one for a cast would look like? (i.e. `fid:<fid>/cast:<hash>`)